cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -398,29 +400,56 @@ var (
 )
 
 func isCoreDNSVersionSupported(client clientset.Interface) (bool, error) {
-	isValidVersion := true
+	var lastError error
+	pollTimeout := 10 * time.Second
+	err := wait.PollImmediate(kubeadmconstants.APICallRetryInterval, pollTimeout, func() (bool, error) {
+		coreDNSPodList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
+			context.TODO(),
+			metav1.ListOptions{
+				LabelSelector: "k8s-app=kube-dns",
+				FieldSelector: "status.phase!=Running",
+			},
+		)
+		if err != nil {
+			lastError = err
+			return false, nil
+		}
+
+		if len(coreDNSPodList.Items) > 0 {
+			lastError = errors.New("find non-running CoreDNS pods")
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	if err != nil {
+		return false, errors.Wrapf(lastError, "after %v", pollTimeout)
+	}
+
 	coreDNSPodList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
 		context.TODO(),
 		metav1.ListOptions{
 			LabelSelector: "k8s-app=kube-dns",
+			FieldSelector: "status.phase=Running",
```

Comment:
list operations are expensive if there are a lot of instances...
why not instead of using a FieldSelector LIST all pods and return an error (continue the poll) if one of them is not Running?